File: avango-connect/python/avango/connect/_io.py
# -*- Mode:Python -*-

##########################################################################
#                                                                        #
# This file is part of Avango.                                           #
#                                                                        #
# Copyright 1997 - 2008 Fraunhofer-Gesellschaft zur Foerderung der       #
# angewandten Forschung (FhG), Munich, Germany.                          #
#                                                                        #
# Avango is free software: you can redistribute it and/or modify         #
# it under the terms of the GNU Lesser General Public License as         #
# published by the Free Software Foundation, version 3.                  #
#                                                                        #
# Avango is distributed in the hope that it will be useful,              #
# but WITHOUT ANY WARRANTY; without even the implied warranty of         #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           #
# GNU General Public License for more details.                           #
#                                                                        #
# You should have received a copy of the GNU Lesser General Public       #
# License along with Avango. If not, see <http://www.gnu.org/licenses/>. #
#                                                                        #
# Avango is a trademark owned by FhG.                                    #
#                                                                        #
##########################################################################

from _registry import _registry_from_type, _registry_from_key

class NoDataAvailable(IndexError):
    pass

def write(name, field, hout):
    field_descriptor = _registry_from_type[type(field)]
    hout.write(name)
    hout.write('\x00')
    hout.write(field_descriptor.key)
    field_descriptor.write(field, hout)
    hout.write('\n')

def read(hin):
    raw_line = hin.readline()
    if raw_line == "":
        raise NoDataAvailable()
    line = raw_line[:-1].split('\x00')
    return line[0], _registry_from_key[line[1]].read(line[2:])

File: attic/avango-shade/core/src/io.cpp
// -*- Mode:C++ -*-

/************************************************************************\
*                                                                        *
* This file is part of AVANGO.                                           *
*                                                                        *
* Copyright 2007 - 2010 Fraunhofer-Gesellschaft zur Foerderung der       *
* angewandten Forschung (FhG), Munich, Germany.                          *
*                                                                        *
* AVANGO is free software: you can redistribute it and/or modify         *
* it under the terms of the GNU Lesser General Public License as         *
* published by the Free Software Foundation, version 3.                  *
*                                                                        *
* AVANGO is distributed in the hope that it will be useful,              *
* but WITHOUT ANY WARRANTY; without even the implied warranty of         *
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the           *
* GNU General Public License for more details.                           *
*                                                                        *
* You should have received a copy of the GNU Lesser General Public       *
* License along with AVANGO. If not, see <http://www.gnu.org/licenses/>. *
*                                                                        *
\************************************************************************/

#include <shade/io.h>
#include <sstream>
#include <boost/tokenizer.hpp>


std::string shade::io::parse_input(std::istream& hin, std::map<std::string, std::string>& keywords)
{
  std::ostringstream contents;
  while (hin.good())
  {
    std::string line;
    std::getline(hin, line);
    contents << line;
    if (line.size()) contents << "\n";

    typedef boost::tokenizer<boost::char_separator<char> > tokenizer;
    boost::char_separator<char> sep(" ");
    tokenizer tokens(line, sep);
    tokenizer::iterator tok_iter = tokens.begin();
    if (    (tok_iter != tokens.end()) && (*(tok_iter++) == "//")
        &&  (tok_iter != tokens.end()) && (*(tok_iter++) == "SHADE") )
    {
      for (; tok_iter != tokens.end(); ++tok_iter)
      {
        std::string::size_type split = tok_iter->find(':');
        std::string key = tok_iter->substr(0, split);
        std::string value;
        if (split != std::string::npos)
        {
          value = tok_iter->substr(split+1, std::string::npos);
        }
        keywords[key] = value;
      }
    }
  }

  return contents.str();
}
